keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,2043 @@
+that the scheduler has deemed suitable. Scheduling the pod is then put on hold
+until all resources are allocated. This avoids scenarios where a Pod is
+permanently assigned to a node which can't fit the pod because of the pod's
+other resource requirements.
+
+When a PodTemplateSpec in an app controller spec references a ResourceClaim by
+name, all Pods created by that controller also use that name and thus share the
+resources allocated for that ResourceClaim. This might be supported by the
+resource driver and can be useful, depending on the characteristics of the
+resource.
+
+But often, each Pod is meant to have exclusive access to its own ResourceClaim
+instance instead. To support such ephemeral resources without having to modify
+all controllers that create Pods, an entry in the new PodSpec.Resources list
+can also be a ResourceClaimTemplate. When a Pod gets created, such a template
+will be used to create a normal ResourceClaim with the Pod as owner, and then
+the normal allocation of the resource takes place.
+
+For immediate allocation, scheduling Pods is simple because the
+resource is already allocated and determines the nodes on which the
+Pod may run. For delayed allocation, a node is selected tentatively
+and driver(s) try to allocate their resources for that node. If that
+succeeds, the Pod can get scheduled. If it fails, the scheduler must
+determine whether some other node fits the requirements and if so,
+request allocation again. If no node fits because some resources were
+already allocated for a node and are only usable there, then those
+resources must be released and then get allocated elsewhere.
+
+This is a summary of the necessary [kube-scheduler changes](#kube-scheduler) in
+pseudo-code:
+
+```
+while <pod needs to be scheduled> {
+  <choose a node, considering potential availability for those resources
+   which are not allocated yet and the hard constraints for those which are>
+  if <no node fits the pod> {
+    if <at least one resource is allocated, unused and was not available on a node> {
+      <randomly pick one of those resources and tell resource driver to free it>
+    }
+  } else if <all resources allocated> {
+    <schedule pod onto node>
+  } else if <some unallocated resource uses delayed allocation> {
+    <tell resource driver to allocate for the chosen node>
+  }
+}
+```
+
+The resources allocated for a ResourceClaim can be shared by multiple
+containers in a pod. Depending on the capabilities defined in the
+ResourceClaim by the driver, a ResourceClaim can be used exclusively
+by one pod at a time, by a certain maximum number of pods, or an
+unlimited number of pods. Kubernetes can only count, but not do more
+than that because it has no understanding of what other limitations might
+apply to specific resources. In practice, one or unlimited are
+expected to be the common cases. A fixed number is supported by the API
+because it is easy to implement and might be
```

Comment:
I agree that "counting users" isn't a required use case and also may require further discussion (do we need to count nodes in addition to pods because there is some kind of attach limit?). So let's drop the "fixed number" part.

But then what should be in the `AllocationResult` instead of `UserLimit int`?

- `UnlimitedUsers bool`? That does not extend well to a fixed number.

- Keep the current `UserLimit int`, but require that it is either 0 or -1 (for unlimited)?

Reply:
```
type Consumption struct {
    Mode ConsumptionMode
     // If we ever have a mode that needs argument, they would go here.
}

type ConsumptionMode string

const (
    ConsumptionModeSingleUser ConsumptionMode = "SingleUser"
    ConsumptionModeMultiUser ConsumptionMode = "MultiUser"
    // One day we might have a "MultiUserCounted" or something
)
```

Reply:
I replaced `UserLimit` with `SharedResource` boolean - `UnlimitedUsers = false` was a double negation.